pallets/rc-migrator/src/accounts.rs

```diff
@@ -0,0 +1,1295 @@
+		match self {
+			AccountState::Part { reserved, .. } => *reserved,
+			AccountState::Migrate => Balance::zero(),
+			AccountState::Preserve => {
+				defensive!("Account must be preserved on RC");
+				Balance::zero()
+			},
+		}
+	}
+	/// Get the consumer count on RC.
+	pub fn get_rc_consumers(&self) -> u32 {
+		match self {
+			AccountState::Part { consumers, .. } => *consumers,
+			// accounts fully migrating to AH will have a consumer count of `0` on Relay Chain since
+			// all holds and freezes are removed.
+			AccountState::Migrate => 0,
+			AccountState::Preserve => {
+				defensive!("Account must be preserved on RC");
+				0
+			},
+		}
+	}
+}
+
+pub type AccountStateFor<T> = AccountState<<T as pallet_balances::Config>::Balance>;
+pub type AccountFor<T> = Account<
+	<T as frame_system::Config>::AccountId,
+	<T as pallet_balances::Config>::Balance,
+	PortableHoldReason,
+	PortableFreezeReason,
+>;
+
+/// Helper struct tracking total balance kept on RC and total migrated.
+#[derive(
+	Encode,
+	DecodeWithMemTracking,
+	Decode,
+	Default,
+	Clone,
+	PartialEq,
+	Eq,
+	RuntimeDebug,
+	TypeInfo,
+	MaxEncodedLen,
+)]
+pub struct MigratedBalances<Balance> {
+	pub kept: Balance,
+	pub migrated: Balance,
+}
+
+pub struct AccountsMigrator<T> {
+	_phantom: sp_std::marker::PhantomData<T>,
+}
+
+impl<T: Config> PalletMigration for AccountsMigrator<T> {
+	type Key = T::AccountId;
+	type Error = Error<T>;
+
+	/// Migrate accounts from RC to AH.
+	///
+	/// Parameters:
+	/// - `last_key` - the last migrated account from RC to AH if any
+	/// - `weight_counter` - the weight meter
+	///
+	/// Result:
+	/// - None - no accounts left to be migrated to AH.
+	/// - Some(maybe_last_key) - the last migrated account from RC to AH if any
+	fn migrate_many(
+		last_key: Option<Self::Key>,
+		weight_counter: &mut WeightMeter,
+	) -> Result<Option<Self::Key>, Error<T>> {
+		// we should not send more than we allocated on AH for the migration.
+		let mut ah_weight = WeightMeter::with_limit(T::MaxAhWeight::get());
+		// accounts batch for the current iteration.
+		let mut batch = XcmBatchAndMeter::new_from_config::<T>();
+
+		let mut iter = if let Some(ref last_key) = last_key {
+			SystemAccount::<T>::iter_from_key(last_key)
+		} else {
+			SystemAccount::<T>::iter()
+		};
+
+		let mut maybe_last_key = last_key;
+		let mut total_items_iterated = 0;
+		loop {
+			// account the weight for migrating a single account on Relay Chain.
+			if weight_counter.try_consume(T::RcWeightInfo::withdraw_account()).is_err() ||
+				weight_counter.try_consume(batch.consume_weight()).is_err()
+			{
+				log::info!(
+					target: LOG_TARGET,
+					"RC weight limit reached at batch length {}, stopping",
+					batch.len()
+				);
+				if batch.is_empty() && total_items_iterated == 0 {
+					defensive!("Not enough weight to migrate a single account");
+					return Err(Error::OutOfWeight);
+				} else {
+					break;
+				}
+			}
+
+			if batch.len() > MAX_ITEMS_PER_BLOCK {
```

Comment:
I have not read all individual pallet migrations yet, but to me it seems that checking `MAX_ITEMS_PER_BLOCK` and `MAX_XCM_MSG_PER_BLOCK` should happen inside the `XcmBatchAndMeter`, so that in all pallets, we check them. I can imagine that maybe in some pallet migrations, we forget to check these.

Reply:
Put differently, these checks should be in batch.push, no?